src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -160,271 +255,252 @@ impl<S: StateStore> SyncedKvLogStoreExecutor<S> {
-                    {
-                        // Apply Vnode Update
-                        write_state.update_vnode_bitmap(&vnode_bitmap);
-                        read_state.update_vnode_bitmap(vnode_bitmap);
-                        first_write_epoch = barrier.epoch;
-                        yield msg;
-                        continue 'recreate_consume_stream;
-                    }
-                    yield msg;
-                }
-            }
-        }
-    }
+            let mut write_future = WriteFuture::receive_from_upstream(input, initial_write_state);
 
-    #[allow(clippy::too_many_arguments)]
-    async fn next(
-        input: &mut BoxedMessageStream,
-        truncation_offset: &mut Option<ReaderTruncationOffsetType>,
-        state_store_stream: &mut Option<StateStoreStream<S>>,
-        flushed_chunk_future: &mut Option<ReadFlushedChunkFuture>,
-        buffer: &mut SyncedLogStoreBuffer,
-        read_state: &mut LogStoreReadState<StateStoreFlushedSnapshotReader<S>>,
-        write_state: &mut LogStoreWriteState<S::Local>,
-        metrics: &mut KvLogStoreMetrics,
-        seq_id: &mut SeqIdType,
-    ) -> StreamExecutorResult<Option<Message>> {
-        select! {
-            biased;
-            // poll from upstream
-            // Prefer this arm to let barrier bypass.
-            upstream_item = input.next() => {
-                match upstream_item {
-                    None => Ok(None),
-                    Some(upstream_item) => {
-                        match upstream_item? {
-                            Message::Barrier(barrier) => {
-                                Self::write_barrier(
-                                    write_state,
-                                    barrier.clone(),
-                                    metrics,
-                                    *truncation_offset,
-                                    seq_id,
-                                    buffer,
-                                ).await?;
-                                Ok(Some(Message::Barrier(barrier)))
+            loop {
+                let select_result = {
+                    let read_future =
+                        read_future.next_chunk(&read_state, &mut buffer, &self.metrics);
+                    pin_mut!(read_future);
+                    let write_future = write_future.next_event();
+                    pin_mut!(write_future);
+                    let output = select(write_future, read_future).await;
+                    drop_either_future(output)
+                };
+                match select_result {
+                    Either::Left(result) => {
+                        // drop the future to ensure that the future must be reset later
+                        drop(write_future);
+                        let (stream, mut write_state, either) = result?;
+                        match either {
+                            Either::Left(message) => {
```

Comment:
Nit: Not very readable to have two `Either::Left`. Prefer to explicitly encode as an enum. e.g. `WriteFutureEvent`.